assets/src/edit-story/app/story/effects/useHistoryEntry.js

```diff
@@ -27,25 +27,34 @@ import { useHistory } from '../../';
 // Record any change to core variables in history (history will know if it's a replay)
 function useHistoryEntry({ story, current, pages, selection, capabilities }) {
   const {
-    actions: { appendToHistory },
+    state: { replayState },
+    actions: { appendToHistory, clearReplayState },
   } = useHistory();
 
   const currentPageIndexRef = useRef();
   const selectedElementIdsRef = useRef();
+  const replayStateRef = useRef();
   useEffect(() => {
     currentPageIndexRef.current = current;
     selectedElementIdsRef.current = selection;
-  }, [current, selection]);
+    replayStateRef.current = replayState;
+  }, [current, selection, replayState]);
 
   useEffect(() => {
-    appendToHistory({
-      story,
-      current: currentPageIndexRef.current,
-      selection: selectedElementIdsRef.current,
-      pages,
-      capabilities,
-    });
-  }, [capabilities, story, pages, appendToHistory]);
+    // If the change was applied by replay, clear the replay flag.
+    if (replayStateRef.current) {
```

Comment:
@barklund @dvoytenko This now keeps the `appendToHistory` true to its name by having a separate action for clearing the `replayState` and doing appending only if it actually appends.

An alternative to this change would be keeping the logic as it was before and instead, rename `ADD_ENTRY` to something more accurate, e.g. `REGISTER_STATE_CHANGE` or something similar -- which reflects more accurately the previous logic.

Let me know your thoughts.

Reply:
I definitely like this way since it's more explicit and less ambiguous w.r.t. add vs replay. However, I'm still not quiet clear w.r.t.

1. Why do we need to have `replayStateRef` for this? E.g. why can't this just be part of the history state `isReplay`. I.e. no matter how time we'd apply the replay - it will always be the same, right?
2. Why do we need `clearReplayState`?

Reply:
1. Using `replayStateRef` to avoid triggering the effect due to the changes of `replayState`. `replayState` is currently like `isReplay` -- if there is an ongoing replay, then `replayState` is truthy, if there isn't, then it's `null`. `replayState` is perhaps not the best naming since it's (in the current implementation) showing the replay pending to be applied, not the static replay that was applied.
The offset reflects the entry that is currently active.
2. `clearReplayState` is used when the replay was already replayed and there is no ongoing replay happening. If we would not clear it out then `appendToHistory` would never be reached since the `replayState` would always be truthy.

Perhaps it would make more sense if `replayState` would be renamed to `replayInProcess` or something similar. Thoughts?

Reply:
`requestedState` could also work. It's like asking for an old revision and getting it returned (but not touching current state) - and then latter in an unrelated event saying "restore state to this old revision, that you gave me".

Reply:
Hm, now replay is a single step process, right? That kind of changes it a lot compared to before.

The reason for the two step process was, that `useHistoryEntry` always just reported the current state to the history provider ignoring whether it was a replay or not. I thought that made the nicest and cleanest api in the history consumer and made "replay or not" an internal detail of the history provider

Now the two things are very tightly coupled and the story provider and history provider knows a lot of details about each other. The separation of concerns is not at all clear anymore, and basically the two things could just straight up merge, as neither can be used without being used **directly** with the other.

Before there were only two interactions between the two:

1. When history says state changes, story changes to whatever history says.
1. When story changes, story reports the change to history.

That's it! That was the entire interface and I actually spend quite some time making such a simple interface work. None of those involve any details about replay, pages, story contents or anything else. There was no interdependency between those, and you could have taken the history provider and apply it to the dashboard or a completely different project and it would still work. Also, you could have exchanged the history provider for some dummy, that only had the current state (if you wanted to save memory on mobile devices by not having history e.g.) and the story provider would work happily without history actually doing anything.

Neither of those things are possible now. The two are tightly coupled and interdependent.

It might work "simpler" at first, but it's a lot more complex to maintain and it's a lot more fragile to changes in the future. If we change how current page works, we suddenly need to change how the history provider works?

I'm not keen on these changes - taking two separate concerns and mashing them together like this is not good software design. IMHO.

Reply:
> Hm, now replay is a single step process, right?

Well, it's still two-step process -- the first step is setting the replay and the second step is clearing the replay -- it's just split up now and the history user has to execute the steps instead of this happening internally.

I agree that it was simpler before -- there was either a request for a replay or a reaction to state changes, history would then know what to do internally. It was definitely convenient that the history users didn't have to be aware of the internal life of the history, whereas now they do need to know when to clear replay and when to add an entry.

It does seem that the original naming needs improving (`ADD_ENTRY` and `replayState` specifically, e.g to `SET_CURRENT_ENTRY` and to `requestEntry` respectively, maybe also `appendToHistory` could be replaced with something like `sendStateChangeToHistory`), and perhaps a design doc would be helpful here, however, the original logic makes more sense to me as well.